src/ReinforcementLearning/SimpleAgent/RandomAgent.h

```diff
@@ -14,26 +15,32 @@ class RandomAgent : public BaseAgent {
     public:
     RandomAgent(){};
     ~RandomAgent(){};
+    std::vector <std::pair<std::vector <bool>, unsigned int>> history;
+    
     /** The member policy selects selectable edges at random.
     */
-    virtual void policy(state* s) const override {
+    virtual int act(state* s) const override {
         std::mt19937 gen(std::chrono::high_resolution_clock::now().time_since_epoch().count());
-        bool hasEdge = true;
-        while (hasEdge) {
-            hasEdge = false;
-            std::vector <int> selectableIndices;
-            for (int j = 0; j < s->selectable.size(); j++) {
-                if (s->selectable[j] == true) {
-                    selectableIndices.push_back(j);
-                    hasEdge = true;
-                }
-            }
-            if (hasEdge == true) {
-                std::uniform_int_distribution<> dis(0, selectableIndices.size() - 1);
-                int d = selectableIndices[dis(gen)];
-                s->select(d);
+        std::vector <int> selectableIndices;
+        bool hasEdge = false;
+        for (int j = 0; j < s->selectable.size(); j++) {
+            if (s->selectable[j] == true) {
+                selectableIndices.push_back(j);
+                hasEdge = true;
             }
         }
+        if (hasEdge == true) {
+            std::uniform_int_distribution<> dis(0, selectableIndices.size() - 1);
+            unsigned int d = selectableIndices[dis(gen)];
+            vector <bool> b1 = s->selectedSubset;
+            std::pair <std::vector <bool>, unsigned int> p1= std::make_pair(b1, d);
+            this->history.push_back(p1);
```

Comment:
This triggers a compiler error as `history` is changed although the act function is declared `const` and is not supposed to change the class members (`history`). Removing `const` from the declaration of act makes the error go away. 
However, it makes sense that the Agent is not changed just because it `act`s once. I suggest you create a class `Policy` with an `act` function that is `const`. The history can be a member of agent, or better: a return value of a function that generates a whole trajectory of actions, states and rewards.

Reply:
Rewards are not included yet, history is a member of Agent and act is now a member of Policy. 
It compiles and - although i haven't made any conclusive tests - at least there is data in history. Hopefully pairs of state and action.